module/zfs/zfs_znode.c

```diff
@@ -822,6 +822,49 @@ zfs_znode_alloc(zfsvfs_t *zfsvfs, dmu_buf_t *db, int blksz,
 
 #else /* APPLE */
 
+#ifdef APPLE_SA_RECOVER
+	/*
+	 * We have had some SA corruption, making for invalid entries. We
+	 * attempt to handle this situation here, by not creating invalid
+	 * type vnodes.
+	 */
+	if (zfs_recover) {
+		if (( IFTOVT((mode_t)mode) == VNON) ||
+			( IFTOVT((mode_t)mode) > VCPLX)) {
+
+			printf("ZFS: WARNING! objid %llu has invalid SA data, please restore from backup. (mode %x)\n",
+				   zp->z_id, (int)zp->z_mode);
+
+			uint64_t parent = zfsvfs->z_recover_parent;
+			if (parent)	{
+
+				zap_cursor_t zc;
+				zap_attribute_t *za;
+				int err;
+				uint64_t mask =  ZFS_DIRENT_OBJ(-1ULL);
+
+				za = kmem_alloc(sizeof (zap_attribute_t), KM_SLEEP);
+				for (zap_cursor_init(&zc, zfsvfs->z_os, parent);
+					 (err = zap_cursor_retrieve(&zc, za)) == 0;
+					 zap_cursor_advance(&zc)) {
+					if ((za->za_first_integer & mask) == (zp->z_id & mask)) {
```

Comment:
It is the mask passed along for comparison, bit AND against the id. As for why you would want to mask bits in the objectID, I wouldn't have a clue. It is already part of zap_search_value() from whence the code came, and I merely set it to the same value the sources always use.